File: src/environment/maze.py
from typing import List, Tuple
from ..config.constants import CellType
from ..config.maze_layouts import MazeSymbols

class Maze:
    def __init__(self, width: int, height: int):
        print(f"Initializing maze with dimensions: {width}x{height}")
        self.width = width
        self.height = height
        self.grid = self._create_empty_maze()
        self.pacman_start = (1, 1)  # Default start position
        self.ghost_starts = []
        
    def _create_empty_maze(self) -> List[List[CellType]]:
        return [[CellType.EMPTY for _ in range(self.width)] 
                for _ in range(self.height)]
    
    def load_layout(self, layout: List[str]):
        """Load maze from a text-based layout"""
        print(f"Loading layout with dimensions: {len(layout[0])}x{len(layout)}")
        self.height = len(layout)
        self.width = len(layout[0])
        self.grid = []
        
        for y, row in enumerate(layout):
            grid_row = []
            for x, cell in enumerate(row):
                if cell == MazeSymbols.WALL:
                    grid_row.append(CellType.WALL)
                elif cell == MazeSymbols.PELLET:
                    grid_row.append(CellType.PELLET)
                elif cell == MazeSymbols.POWER_PELLET:
                    grid_row.append(CellType.POWER_PELLET)
                elif cell == MazeSymbols.PACMAN_START:
                    self.pacman_start = (x, y)
                    grid_row.append(CellType.PATH)
                elif cell == MazeSymbols.GHOST_START:
                    self.ghost_starts.append((x, y))
                    grid_row.append(CellType.PATH)
                else:
                    grid_row.append(CellType.PATH)
            self.grid.append(grid_row)
            
        print(f"Final grid dimensions: {len(self.grid)}x{len(self.grid[0])}")

    def get_cell_type(self, x: int, y: int) -> CellType:
        """Get the type of cell at the given position"""
        if 0 <= y < self.height and 0 <= x < self.width:
            return self.grid[y][x]
        return CellType.WALL

    def set_cell_type(self, x: int, y: int, cell_type: CellType):
        """Set the type of cell at the given position"""
        if 0 <= y < self.height and 0 <= x < self.width:
            self.grid[y][x] = cell_type

    def is_valid_position(self, x: int, y: int) -> bool:
        """Check if a position is valid and not a wall"""
        return (0 <= x < self.width and 
                0 <= y < self.height and 
                self.grid[y][x] != CellType.WALL)
    
    def eat_pellet(self, x: int, y: int) -> bool:
        """Eat a pellet at the given position and return True if it was a power pellet"""
        if 0 <= x < self.width and 0 <= y < self.height:
            cell_type = self.grid[y][x]
            if cell_type in [CellType.PELLET, CellType.POWER_PELLET]:
                is_power_pellet = cell_type == CellType.POWER_PELLET
                self.grid[y][x] = CellType.PATH
                return is_power_pellet
        return False
    
    def count_remaining_pellets(self) -> int:
        """Count the number of remaining pellets in the maze"""
        count = 0
        for y in range(self.height):
            for x in range(self.width):
                if self.grid[y][x] in [CellType.PELLET, CellType.POWER_PELLET]:
                    count += 1
        return count

File: web/js/core/maze.js
import { CellType } from '../config/constants.js';
import { MazeSymbols } from '../config/mazelayouts.js';

export class Maze {
    constructor(width, height) {
        console.log(`Initializing maze with dimensions: ${width}x${height}`);
        this.width = width;
        this.height = height;
        this.grid = this._createEmptyMaze();
        this.pacmanStart = { x: 1, y: 1 };
        this.ghostStarts = [];
    }

    _createEmptyMaze() {
        return Array(this.height).fill()
            .map(() => Array(this.width).fill(CellType.EMPTY));
    }

    loadLayout(layout) {
        console.log(`Loading layout with dimensions: ${layout[0].length}x${layout.length}`);
        this.height = layout.length;
        this.width = layout[0].length;
        this.grid = [];
        
        for (let y = 0; y < layout.length; y++) {
            const gridRow = [];
            for (let x = 0; x < layout[y].length; x++) {
                const cell = layout[y][x];
                switch (cell) {
                    case MazeSymbols.WALL:
                        gridRow.push(CellType.WALL);
                        break;
                    case MazeSymbols.PELLET:
                        gridRow.push(CellType.PELLET);
                        break;
                    case MazeSymbols.POWER_PELLET:
                        gridRow.push(CellType.POWER_PELLET);
                        break;
                    case MazeSymbols.PACMAN_START:
                        this.pacmanStart = { x, y };
                        gridRow.push(CellType.PATH);
                        break;
                    case MazeSymbols.GHOST_START:
                        this.ghostStarts.push({ x, y });
                        gridRow.push(CellType.PATH);
                        break;
                    default:
                        gridRow.push(CellType.PATH);
                }
            }
            this.grid.push(gridRow);
        }
        
        console.log(`Final grid dimensions: ${this.grid.length}x${this.grid[0].length}`);
    }

    getCellType(x, y) {
        if (y >= 0 && y < this.height && x >= 0 && x < this.width) {
            return this.grid[y][x];
        }
        return CellType.WALL;
    }

    setCellType(x, y, cellType) {
        if (y >= 0 && y < this.height && x >= 0 && x < this.width) {
            this.grid[y][x] = cellType;
        }
    }

    isValidPosition(x, y) {
        return (x >= 0 && x < this.width && 
                y >= 0 && y < this.height && 
                this.grid[y][x] !== CellType.WALL);
    }

    eatPellet(x, y) {
        if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
            const cellType = this.grid[y][x];
            if (cellType === CellType.PELLET || cellType === CellType.POWER_PELLET) {
                const isPowerPellet = cellType === CellType.POWER_PELLET;
                this.grid[y][x] = CellType.PATH;
                return isPowerPellet;
            }
        }
        return false;
    }

    countRemainingPellets() {
        let count = 0;
        for (let y = 0; y < this.height; y++) {
            for (let x = 0; x < this.width; x++) {
                if (this.grid[y][x] === CellType.PELLET || 
                    this.grid[y][x] === CellType.POWER_PELLET) {
                    count++;
                }
            }
        }
        return count;
    }
    
    draw(ctx, offset = 0) {
        const cellSize = 30; // You might want to move this to constants
        
        for (let y = 0; y < this.height; y++) {
            for (let x = 0; x < this.width; x++) {
                const cellType = this.grid[y][x];
                const pixelX = x * cellSize;
                const pixelY = y * cellSize + offset;
                
                switch (cellType) {
                    case CellType.WALL:
                        ctx.fillStyle = 'blue';
                        ctx.fillRect(pixelX, pixelY, cellSize, cellSize);
                        break;
                    case CellType.PELLET:
                        ctx.fillStyle = 'white';
                        ctx.beginPath();
                        ctx.arc(
                            pixelX + cellSize/2,
                            pixelY + cellSize/2,
                            2,
                            0,
                            Math.PI * 2
                        );
                        ctx.fill();
                        break;
                    case CellType.POWER_PELLET:
                        ctx.fillStyle = 'white';
                        ctx.beginPath();
                        ctx.arc(
                            pixelX + cellSize/2,
                            pixelY + cellSize/2,
                            6,
                            0,
                            Math.PI * 2
                        );
                        ctx.fill();
                        break;
                }
            }
        }
    }
}